ui/objects/semantic-ui/js/objects/editform/components/VocabularyTreeSelectField.jsx

```diff
@@ -0,0 +1,428 @@
+
+export const VocabularyTreeSelectField = ({
+  fieldPath,
+  multiple,
+  optionsListName,
+  usedOptions = [],
+  helpText,
+  optimized,
+  ...uiProps
+}) => {
+  const { formConfig } = useFormConfig();
+  const { vocabularies } = formConfig;
+  const formik = useFormikContext();
+  const { values, setFieldTouched } = useFormikContext();
+  const value = getIn(values, fieldPath, multiple ? [] : {});
+
+  const { all: allOptions, featured: featuredOptions } =
+    vocabularies[optionsListName];
+
+  if (!allOptions) {
+    console.error(
+      `Do not have options for ${optionsListName} inside:`,
+      vocabularies
+    );
+  }
+
+  const [openState, setOpenState] = useState(false);
+  const [parentsState, setParentsState] = useState([]);
+  const [keybState, setKeybState] = useState([]);
+  const [selectedState, setSelectedState] = useState([]);
+
+  const serializedOptions = useMemo(
+    () => serializedVocabularyItems(allOptions),
+    [allOptions]
+  );
+
+  const handleOpen = () => {
+    setOpenState(true);
+  };
+
+  const hierarchicalData = useMemo(() => {
+    let data = [];
+    let currentRow = [];
+    let currentLevel = 0;
+
+    serializedOptions.forEach((option, index) => {
+      if (option.hierarchy.ancestors.length === currentLevel) {
+        currentRow.push(option);
+      } else {
+        data.push(currentRow);
+        currentRow = [option];
+        currentLevel++;
+      }
+    });
+
+    if (currentRow.length > 0) {
+      data.push(currentRow);
+    }
+
+    return data;
+  }, [serializedOptions]);
+
+  const amount = hierarchicalData.length;
+
+  const updateHierarchy = (parent, index) => () => {
+    let updatedParents = [...parentsState];
+    updatedParents.splice(index + 1);
+    updatedParents[index] = parent;
+
+    let updatedKeybState = [...keybState];
+
+    const columnOptions = hierarchicalData[index];
+    const nextColumnIndex = columnOptions.findIndex((o) => o.value === parent);
+    updatedKeybState.splice(index + 1);
+    updatedKeybState[index] = nextColumnIndex;
+
+    setParentsState(updatedParents);
+    setKeybState(updatedKeybState);
+  };
+
+  const handleSelect = (option, val, e) => {
+    e.preventDefault();
+    const existingIndex = selectedState.findIndex(
+      (i) => i.value === option.value
+    );
+
+    if (existingIndex !== -1) {
+      setSelectedState((prevState) => {
+        const newState = [...prevState];
+        newState.splice(existingIndex, 1);
+        return newState;
+      });
+    } else {
+      if (multiple && selectedState.length != 0) {
+        setSelectedState((prevState) => [...prevState, option]);
+      } else {
+        setSelectedState([option]);
+      }
+    }
+  };
+
+
+  const handleSubmit = () => {
+    let prepSelect;
+    if (multiple) {
+      prepSelect = selectedState.map((i) => serializeVocabularyItem(i.value));
+      const existingValues = getIn(formik.values, fieldPath, []);
+      prepSelect = [...existingValues, ...prepSelect];
```

Comment:
What happens, when some value is both in selectedState & in existingValues -- wouldn't it get duplicated?